### ISSUE TITLE
fast forward: set next block timestamp in the future

### ISSUE THREAD

**Issue by gsmachado** [neo-project/neo-express#207]
As mentioned in https://github.com/neo-project/neo-express/issues/206#issuecomment-1050828900, having a way to set the next block timestamp at any point in the future would cover (IMHO) 99% of tests in time-dependent smart contracts. This would be more relevant in the online mode (but maybe there are use cases for the offline one as well...).

Tools such as hardhat/ganache provide a way to do that and it's good enough for the majority of cases.

**Comment by gsmachado** [comment]
@devhawk any idea if this one could be prioritized in the near future? 😄 

This would avoid us (and other smart contract devs) to do [this](https://github.com/AxLabs/grantshares-contracts/blob/proposal-expiry/src/main/java/com/axlabs/neo/grantshares/GrantSharesGov.java#L324-L325). Because now we don't have a way to set the next block timestamp in the future, and we want to stick with our smart contract logic.

**Comment by devhawk** [comment]
I'll take a look at this today

**Comment by devhawk** [comment]
How are you thinking this feature would surface on the command line? I'm thinking that we would add an option for the developer to specify the timestamp of the last block generated as part of the fast forward. But I'm not sure how the timestamp should be specified. 

* Delta in seconds from the current time
* Delta in seconds from latest block
* Explicitly specifying the final timestamp in some to be decided on string format
* something else?

**Comment by gsmachado** [comment]
> How are you thinking this feature would surface on the command line? I'm thinking that we would add an option for the developer to specify the timestamp of the last block generated as part of the fast forward. But I'm not sure how the timestamp should be specified.
> 
> * Delta in seconds from the current time
> * Delta in seconds from latest block
> * Explicitly specifying the final timestamp in some to be decided on string format
> * something else?

We believe that the option "Delta in seconds from latest block" is the most intuitive for developers.

So, let's clear something out here:

We assume you mean that the NEXT block (after we run the fast fwd command) will have the specified timestamp, right? And then, the blocks generated after the "fast forward with timestamp command" should have timestamps in the future -- and not based on local time. This is important.

**Comment by devhawk** [comment]
> We assume you mean that the NEXT block (after we run the fast fwd command) will have the specified timestamp, right? And then, the blocks generated after the "fast forward with timestamp command" should have timestamps in the future -- and not based on local time. This is important.

I was suggesting that the LAST block generated by `fastfwd` would have the specified time. The NEXT block generated after running `fastfwd` would be 1 ms after the time in the `fastfwd` command. Every block after that would also only increment the timestamp by 1ms until DateTime.UtcNow catches up. Does that work?

Also, I was planning on spreading the generated blocks out evenly over the period specified by `fastfwd` so if you generate 100 blocks but specify a timestamp 10,000 seconds in the future, there would be roughly 100 seconds between timestamps for the generated blocks. The spacing will not be guaranteed to be exactly even, but I would make a best effort. Does that work?

**Comment by gsmachado** [comment]
oh... @csmuller and I tried to interpret [your previous comment](https://github.com/neo-project/neo-express/issues/207#issuecomment-1077934209) and we concluded: "hmm, I think devhawk doesn't actually mean that the LAST block generated by `fastfwd` would have the specified timestamp.... because this would create a huge amount of blocks for absolutely nothing.... so, I think he meant the NEXT block.. but let's ask and confirm"

But, actually, that's what you meant.

> Also, I was planning on spreading the generated blocks out evenly over the period specified by fastfwd so if you generate 100 blocks but specify a timestamp 10,000 seconds in the future, there would be roughly 100 seconds between timestamps for the generated blocks. The spacing will not be guaranteed to be exactly even, but I would make a best effort. Does that work?

Then, based on what you've suggested above ☝️ , which in fact is a bit more powerful than what I initially planned, let's imagine the following scenario.

We would need to fast forward to a block where the (delta) timestamp is 1 year from the last block -- which is a reasonable scenario for a time-sensitive smart contract. So, `86400` sec per day * `365` = `31536000` sec per year.

We would do, something like this:

```shell
neoxp fastfwd --timestamp-delta-last-block 31536000 1
```

Considering that the LAST block timestamp (at the moment before the `neoxp fastfwd` cmd) was `27 Mar, 2022, 3:00:00 PM` and block index `1`, then the NEXT block would have the timestamp set to `27 Mar, 2023, 3:00:00 PM` and block index `2`. Right?

Now, let's imagine the same scenario but executing the following:

```shell
neoxp fastfwd --timestamp-delta-last-block 31536000 5
```

Then, considering that the last block timestamp (at the moment before the `neoxp fastfwd` cmd) was `27 Mar, 2022, 3:00:00 PM` and block index `1`, then the next blocks would be:
- block `2`: `8 Jun 2022, 3:00:00 PM`
- block `3`: `20 Aug 2022, 3:00:00 PM`
- block `5`: `1 Nov 2022, 3:00:00 PM`
- block `6`: `13 Jan 2023, 3:00:00 PM`
- block `7`: `27 Mar 2023, 3:00:00 PM` (`6307200` sec interval between blocks -- not guaranteed, though, best effort)
- block `8`: `27 Mar 2023, 3:00:01 PM` (considering that `neoxp` was started with `--seconds-per-block 1`)
- block `9`: `27 Mar 2023, 3:00:02 PM`

Right?

Also, @csmuller, do you think that is this fine for our SC tests for time-sensitive smart contracts?

**Comment by devhawk** [comment]
Is there a standard text format for time deltas? C# has a [set of characters] (https://docs.microsoft.com/en-us/dotnet/standard/base-types/custom-date-and-time-format-strings) for representing different segments of date time. Might be nice to be able to specify `1y` instead of calculating the number of seconds in a year.

On the other hand, not sure this feature will be used enough to warrant that

**Comment by devhawk** [comment]
We might want to have a quick chat on discord. I didnt realize that i was proposing something more powerful than you were asking for

**Comment by gsmachado** [comment]
> We might want to have a quick chat on discord. I didnt realize that i was proposing something more powerful than you were asking for

Oh... the only thing that is more "powerful" than originally thought is your proposal to be able to combine `fastfwd` of blocks together with the final timestamp (last block). Just that.

Initially, we thought that having a feature just to set the timestamp of the next block was totally sufficient to cover our time-sensitive smart contracts.

However, thinking a bit better, I believe it totally makes sense to combine things... giving users the freedom to set how many blocks to fast forward, and also set the last timestamp. Because setting the timestamp of the next block would be achievable with something like `neoxp fastfwd --timestamp-delta-last-block 31536000 1`, which is great.

**Comment by gsmachado** [comment]
> Is there a standard text format for time deltas? C# has a [set of characters] (https://docs.microsoft.com/en-us/dotnet/standard/base-types/custom-date-and-time-format-strings) for representing different segments of date time. Might be nice to be able to specify `1y` instead of calculating the number of seconds in a year.
> 
> On the other hand, not sure this feature will be used enough to warrant that

Yeah, that would be fine, I think.

However, as far as I read, it would be a bit difficult to specify something like `31536005` seconds (1 year and 5 seconds) if developers need it. Maybe we should support the [standard you proposed](https://docs.microsoft.com/en-us/dotnet/standard/base-types/custom-date-and-time-format-strings) and also the raw format in seconds (if no suffix char is specified)?

**Comment by csmuller** [comment]
About the time format: Since time in smart contracts will probably be specified in milliseconds I don't think it's bad if the command only allows time in milliseconds. Having a more human readable format could be added in a later step/issue. But if you want to tackle everything in one go then I agree to a format like y (years), d (days), h (hours), s (seconds).

**Comment by devhawk** [comment]
Using [C# custom timespan format strings](https://docs.microsoft.com/en-us/dotnet/standard/base-types/custom-timespan-format-strings) as a guide, we would have the fairly obvious:

* `d` for day
* `h` for hour
* `m` for minute
* `s` for second

However, do we also need timespans longer than a day? `y` for year and `w` for week are obvious choices. But what about month? `m` is already taken for minute. 

Also, years and months don't have specific lengths. I do not think we want to get into any kind of accurate month/year length calculation. So maybe we just stick with the above? or above + `w` for week? `52w` for year is much clearer than `3153600`

**Comment by devhawk** [comment]
FYI, the C# `TimeSpan` class already has an existing text format:

`d.hh:mm[:ss[.ff]] | hh:mm[:ss[.ff]]`

> note, `d`, `h`, `m` and `s` are as discussed above. `f` stands for "fractional seconds`

Using the existing `TimeSpan.TryParse` method is much easier than building my own time span parser, so I'm just going to go w/ this format with the following logic.

* If the timestamp delta value can be parsed as an unsigned long, that value will be used as the timestamp delta in seconds
* If the timestamp delta value can be parsed as a `TimeSpan`, use the `TimeSpan.TotalMilliseconds` property as the delta
* if the parsed timestamp is negative or can't be parsed as a `ulong` or `TimeSpan`, throw exception

**Comment by gsmachado** [comment]
> FYI, the C# `TimeSpan` class already has an existing text format:
> 
> `d | d.hh:mm[:ss[.ff]] | hh:mm[:ss[.ff]]`
> 
> > note, `d`, `h`, `m` and `s` are as discussed above. `f` stands for "fractional seconds`
> 
> I thought it better to use the existing `TimeSpan.TryParse` method than to roll my own time span parser. I'm thinking this would be the logic:
> 
> * If the timestamp delta value can be parsed as an unsigned long, that value will be used as the timestamp delta in seconds
> * If the timestamp delta value can be parsed as a `TimeSpan`, use the `TimeSpan.TotalMilliseconds` property as the delta
> * if the parsed timestamp is negative or can't be parsed as a `ulong` or `TimeSpan`, throw exception
> 
> This means that to specify a time in days, you need to pass a string like `1.0:0`. The `.0:0` suffix is required. I considered simply parsing as a TimeSpan, but a single integer value is treated as days, which seems like a poor choice for this command.

I believe it's fine as you suggested, with the use of C# `TimeSpan`. The important thing is to be flexible and give the possibility to specify in days or seconds (or even in milliseconds) -- which is there.

The only drawback, as you said, is the use of a single digit being interpreted as days, and not seconds (which would be the preferable choice). But I'm sure devs live with that. 😄 

@csmuller ay strong opinions on this? or ready to go?

**Comment by devhawk** [comment]
> The only drawback, as you said, is the use of a single digit being interpreted as days, and not seconds (which would be the preferable choice). But I'm sure devs live with that. 😄

No, I first try to parse the value as an unsigned long. If successful, it's treated as seconds. Only if ulong.TryParse fails do I try to parse it as a TimeSpan.

Here's the actual code:

``` cs
internal static TimeSpan ParseTimestampDelta(string timestampDelta)
    => string.IsNullOrEmpty(timestampDelta)
        ? TimeSpan.Zero
        : ulong.TryParse(timestampDelta, out var @ulong)
            ? TimeSpan.FromSeconds(@ulong)
            : TimeSpan.TryParse(timestampDelta, out var timeSpan)
                ? timeSpan
                : throw new Exception($"Could not parse timestamp delta {timestampDelta}");
```

**Comment by devhawk** [comment]
fixed by #224
